[P118] Make CS pin configurable, some refactoring

tonhuisman opened the pull request:
Features/improvements:
- [x] Make the `CS` pin configurable
- [x] Resolve some compiler warnings
- [x] Leave SPI initialization to ESPEasy core
- [x] Refactor to use `Plugin_data_struct` to enable multi-instance and save some memory
- [x] Format source code using Uncrustify
- [x] Add support for Orcon devices, taken (with permission) from #4099 by @nl0pvm 
- [x] ~Moved to `Collection E` because of the size increase.~
- [x] Add `CLIMATE` builds and move this plugin there

TODO:
- [x] Test on ESP8266 ([confirmed](https://github.com/letscontrolit/ESPEasy/pull/3889#issuecomment-1211759295))
- [x] Test on ESP32
- [x] Restore plugin in ESP32 builds
- [x] Test Orcon commands (help requested) ([confirmed](https://github.com/letscontrolit/ESPEasy/pull/3889#issuecomment-1254990280))

tonhuisman committed:
[P118] Make CS pin configurable, resolve some compiler warnings

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

svollebregt commented:
Shall I already test this on my ESP8266+CC1101 or are you planning to add the Plugin_data_struct soon?

tonhuisman commented:
> Shall I already test this on my ESP8266+CC1101 or are you planning to add the Plugin_data_struct soon?

I'm close to committing those changes as well, not sure I'll manage that today, could be tomorrow end of morning.

tonhuisman committed:
[P118] Restructure using Plugin_data_struct, allow 3 simultaneous instances, code reformatted

tonhuisman committed:
[P118] Add library.properties and reference it

TD-er commented on lib/Itho/CC1101.h:
```diff
@@ -168,7 +168,7 @@ class CC1101
 	//functions
 	public:
-		CC1101();
+		CC1101(int8_t CSpin = SS);
```
Where is this awfully chosen define or global variable `SS` defined?
If it is in our code, let's rename it and if it is elsewhere, let's make a PR to change it.

tonhuisman replied:
AFAICS, it is an Arduino defined constant global, there are also the MISO, MOSI, etc. and D0, D1 constants. Not sure that will ever change 🤔

TD-er replied:
Hmm OK, that will not likely change indeed. (the same as SuperSpeed USB3, which even worse as it often also has the charging symbol next to it)

tonhuisman committed:
[P118] Use more descriptive names for a few variables/constants

tonhuisman committed:
[P118] Fix missing #define

tonhuisman committed:
[Build] Fix compiler warnings

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -124,10 +124,10 @@ boolean Plugin_118(uint8_t function, struct EventStruct *event, String& string)
-      PIN(0)     = -1;        // Interrupt pin undefined by default
-      PIN(1)     = 15;        // CS pin use the previous default of SS/gpio 15
+      PIN(0)     = -1;         // Interrupt pin undefined by default
+      PIN(1)     = PIN_SPI_SS; // CS pin use the previous default of PIN_SPI_SS/gpio 15
```
The GPIO comment is only ESP8266

tonhuisman replied:
I can add a conditional compile for ESP32 to set it to -1 by default. I would expect the default Slave Select to be the same for ESP32, when using the 'default' SPI pins. But I'd expect the user to select the desired CS pin anyway. Then 15 isn't a bad default, but also -1 wouldn't be wrong, just require them to pick one. Same situation for the interrupt pin.
The conditional definition for this value is ~stolen~borrowed from the Arduino source, so it works for ESP32 too.

TD-er replied:
What I meant is that GPIO-15 is ESP8266 only, so a simple `(on ESP8266)` would be enough here in the comment line ;)
But indeed, had not thought of the default needed for ESP32.
If it is already a sensible value in the Arduino source then you don't need to do the #ifdef stuff.

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -236,7 +236,7 @@ boolean Plugin_118(uint8_t function, struct EventStruct *event, String& string)
-	  // Find matching interrupt flag
+      // Find matching interrupt flag
       for (uint8_t i = 0; i < P118_INTERUPT_HANDLER_COUNT && offset == -1; i++) {
         if (PLUGIN_118_Task[i] == event->TaskIndex) {
```
Is this section still needed when using the datastruct setup?
Or is it needed for the interrupthandling? (thus volatile)

Maybe also have a look at the pulse counter plugin. P003.

tonhuisman replied:
Ah, I tried to avoid the previous implementation limit for P003 being restricted to tasks 1..4, but that seems to be lifted by the current implementation. I can do a similar implementation, having no explicit limits, other than the number of pins supporting interrupts, just give me a little time 😃.

tonhuisman replied:
Refactoring actually went smoother than first estimated 😉

tonhuisman committed:
[P118] Refactor interrupt handling

svollebregt commented:
I downloaded the branch from @tonhuisman and build it using Atom. However, when I flash my ESP either OTA or using the ESP Easy Flasher my device will not boot anymore.

Below the output of the serial log:
`;d d£ƒ| îlα|♥♀♦♦î♦d∞♦#<╟├♥Σ↕█{Æ#ä♀c─≤on▀lo'▄πΣ♦#∟8Σ╟$;dsl8√'α►♥♦♦é♦d♀─£♀♀♦c♀nΓ|♥Σdç♦╟c─√go∩ lîÅd`♥╪←←'o♦d`♥♥gsçôÆn♦♀#♥l`♥☼s██'♀♦c♥l`♥£#♦♦ç∟♥¢l;¢`♥ⁿén▄♥¬U83 : Info   :

INIT : Booting version: pygit2_not_installed (ESP82xx Core 2843a5ac, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2 PUYA support)
84 : Info   : INIT : Free RAM:31872
85 : Info   : INIT : Manual Reboot #10 - Restart Reason: External System
86 : Info   : FS   : Mounting...
111 : Info`

tonhuisman commented:
Hmm, there is always a risk when refactoring that something breaks 🤔 
If you download the Github Actions build [from here](https://github.com/letscontrolit/ESPEasy/actions/runs/1630548565), and extract a TEST_D build that matches your ESP, does that work any better?
Flashing a working build, then disabling the P118 task before upgrading, and then do the upgrade should at least get you a running unit. Then from Tools page you could execute `taskenable,<tasknumber>` to try it (but it might still fail/crash, but you won't go into a boot-loop as the task is then still disabled at boot).

svollebregt commented:
First disabling P118 and then flashing it with that build did result in a successful boot. But it immediately crashes when I enable the task associated with P118. In the weblog which I had open in another tab nothing is visible.

tonhuisman commented:
Any crashlog is only really visible in the serial log, the weblog can't be updated anymore if the unit is crashing.
So I have some investigation to do. If you could get that crashlog, that would be very helpful.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Attempt to avoid startup crash

tonhuisman commented:
@svollebregt I've tried to move some bits, and reverted the disable/enable interrupt logic I accidentally introduced when moving the code to `Plugin_data_struct`.
If, during testing, you could watch the serial log, then you'd have a better look at the actual crash information (both before and after the big hex-dump it usually outputs), if that still happens, then please post those parts of the log here.
NB: Download [from here](https://github.com/letscontrolit/ESPEasy/actions/runs/1650753062), for this updated build.

svollebregt commented:
I connected the device to my PC and opened a serial monitor with your latest commit loaded, with the serial output set to Debug Dev. When I enable the plugin I see the following messages:

> 152783 : Info   : HTTP: taskenable,2
152787 : Debug  : Command: taskenable
152791 : Info   : INIT PLUGIN_118
152792 : Info   : Extra Settings PLUGIN_118 loaded

No crash dump is generated. I tried this using the serial monitor of Atom, Arduino IDE, and MobaXterm. Do I need a special tool to see a crash log, or does the ESP simply not crash but just freezes?

TD-er commented:
To decode the crash dump, you need the ESP connected to the IDE (VS code, not sure if Atom allows the stackdecoder to run) and run "monitor" from the PlatformIO menu.
You also need to have the code loaded and the right .elf file, so it is best to compile it and upload it from your PC to make sure you have the correct .elf file related to the code you're running.

tonhuisman commented:
The $64000,- question: _Does_ it still crash? (I can't determine that from your comment...)

TD-er commented:
> The $64000,- question: _Does_ it still crash? (I can't determine that from your comment...)

I wouldn't be surprised if it did...
See also why I disabled it [here](https://github.com/letscontrolit/ESPEasy/pull/3845/commits/38cd32e0d7e1ad102bd94e1a7b0407f547844ad3) (temporary) for the IDF4.4 PR as just having it present in that build will cause crashes/boot loops on the ESP32.
There is something fishy with the callback functions (or attached to interrupts)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman commented:
Well, there is another observation: If the plugin is started when there is no hardware connected, it just hangs forever. Not even the WDT kicks in to revive the unit. At least that's what I get when enabling the plugin, and that's exactly the same behavior I got _before_ I started the refactoring.
Not sure if @svollebregt is testing with the hardware connected or not, but this plugin won't work without it. (We might need to note that somewhere).

tonhuisman commented:
And this is the culprit: https://github.com/letscontrolit/ESPEasy/blob/mega/lib/Itho/CC1101.cpp#L31-L34
That explains why the WDT isn't kicking in.
It is called on each `writeCommand()` and `writeRegister()` call, and a few other places...

svollebregt commented:
I was testing it with all hardware attached (minus a CO2 sensor as it uses the serial port, disabled that plugin). I switched to VS Studio, and performed a successful build and monitor. Output on the terminal is the same, seems it hangs and does not crash.

tonhuisman commented:
I see a couple of while loops (at least in the CC1101 library) that have the potential to last 'quite long' (as in 'forever'), if an expected state isn't reached.
I've tried to fix the blocking loop I referenced, but it is not enough. Without the hardware available it will be quite hard to debug, so I'll order such a board to dive deeper into this. Until that arrives, this PR will go on hold, unless of course you are able to the debugging 😉 
I've also changed the declaration of the interrupt handler, according to another example, where the `ICACHE_RAM_ATTR` attribute is applied to the function declaration, and _not_ to the implementation. I'll commit that change in a minute.

tonhuisman committed:
[P118] Attempt to avoid startup crash

svollebregt commented:
I will test your updated code later today, will keep you posted.

TD-er commented:
> And this is the culprit: https://github.com/letscontrolit/ESPEasy/blob/mega/lib/Itho/CC1101.cpp#L31-L34

Here is another one: https://github.com/letscontrolit/ESPEasy/blob/41b58b2a2de0e63e038a9239aac33e8e29b89d86/lib/Itho/CC1101.cpp#L125-L130

TD-er commented:
And this one is quite tricky as there is just a semicolon, not an empty scope (`{}`)
Well it is potentially a wait forever loop.

https://github.com/letscontrolit/ESPEasy/blob/41b58b2a2de0e63e038a9239aac33e8e29b89d86/lib/Itho/CC1101.cpp#L255-L258

TD-er commented:
Please run uncrustify on that .cpp file, just to see what else may be tricky like the lines I mentioned above.
Strictly speaking you don't need to add curly braces if you only need to do a single line after an if, while, etc.  but it may lead to errors very easily if you don't.

tonhuisman committed:
[P118] Add entry to Custom-sample.h

svollebregt commented:
Update: same behaviour after compiling your last commit (eeb655d), it just hangs after the extra settings Plugin_118 loaded statement, no crash dump.

tonhuisman commented:
The most changes Uncrustify did were adjustments to the white-space and also added a couple of missing scope-braces, like the `while(...);` with `while(...) {};`
Maybe I could add a limit to the number of loops that is done per operation, but I have no idea how long a specific response could take to be valid.
I ordered a CC1101 board, but it will take at least a couple of weeks before that arrives. As the current software is working as intended, no need to rush this, I think.

tonhuisman committed:
[P118] Reformat CC1101 library using Uncrustify

TD-er commented:
Most are SPI commands, so I doubt you will ever need more than 1 sec to get a reply.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Log optimization

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Code optimizations

svollebregt commented:
@tonhuisman What is the current status of the attempts to change the code of P118 to allow multiple instances and prevent a start-up freeze when it cannot find the CC1101? I noticed you made some changes in your own repository. I have been idle for a while due to having zero free time, but I should be able to do some testing again in the coming period.

I saw in the release notes P118 is currently disabled as it crashed with IDF4.4? Is this related to the refactoring or an inherent property of the plugin/library code?

tonhuisman commented:
I have received the hardware some time ago, _and_ have installed an Itho ventilation unit to replace the previous, different brand broken down, one, I just need to connect the CC1101 board to an ESP and test it myself... I've been distracted by many other things going on 🤔
The IDF 4.4 issue was detected & handled by TD-er, actual testing should reveal any issues there, so will have to see how that works out when I set it up.

TD-er commented:
> I saw in the release notes P118 is currently disabled as it crashed with IDF4.4? Is this related to the refactoring or an inherent property of the plugin/library code?

I don't have a clue why, but when this plugin was included in a 16M build, it was continuously rebooting.
On 4M builds it was working fine.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Minor code optimizations

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Itho library limit formerly endless loops to max. 3 seconds

tonhuisman committed:
[P118] Fix bugs found during testing, rename variables, clean up source

tonhuisman committed:
[P118] Update documentation

tonhuisman committed:
[P118] Fix bugs found during testing, rename variables, clean up source

tonhuisman commented:
@svollebregt I finally got the courage to wire up the CC1101 and start testing my changes. I fixed a couple of glorious bugs (as was to be expected 😉), cleaned up the source some more and updated the documentation.

Would you care to test this, please? (I had to replace my home-ventilation unit earlier this year, so chose an ITHO 😄, it does work as intended for me.)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

svollebregt commented:
Works flawless on my side, great work! Would it be possible to make the information in the Devices overview tab less ambiguous? The label CS is indicating the CC1101 interrupt pin on my ESP, while the CS pin has no label:

CLK: GPIO-14 (D5)
MISO: GPIO-12 (D6)
MOSI: GPIO-13 (D7)
CS: GPIO-4 (D2)
GPIO-15 (D8)

tonhuisman commented:
> make the information in the Devices overview tab less ambiguous?

That is a generic issue for SPI devices, I'll see if I can fix that.

And the reason it is showing the labels 'swapped' is that I kept the `PIN(n)` configuration as is, for backward compatibility, but the default is to have `CS` in `PIN(0)`.

tonhuisman commented:
> I'll see if I can fix that.

![image](https://user-images.githubusercontent.com/1922393/184141842-1e1aed7d-d193-404a-a282-1fb038a5c3ab.png)

(This is on ESP32, so no `Dx` names)

Wait for the next commit to test on ESP32 (and -S2), as I've just fixed that but not yet pushed.

tonhuisman committed:
[P118] Fix ESP32 support, shorten lib time-out, improve Devices overview display

tonhuisman committed:
[Devices] Fine-tune display of custom GPIO configuration

tonhuisman committed:
[P116] Add Devices list GPIO overview

tonhuisman committed:
[P118] [Build] Restore Itho communication in ESP32 builds

tonhuisman committed:
[Build] Remove NeoPixel plugins from DISPLAY builds

(cherry picked from commit 6074be2397148861d3a02cb055d8e6a32694aa3b)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[Itho Lib] Merge Orcon functions from #4099

tonhuisman committed:
[P118] Merge Orcon functions from #4099

tonhuisman committed:
[P118] Update documentation with Orcon commands

tonhuisman commented:
The build is failing because of the increased size. I may need to move this plugin to `Collection E` to make builds fit again.

tonhuisman committed:
[P118] Move plugin to `Collection E` to make the build fit in again

TD-er commented on lib/Itho/IthoCC1101.cpp:
```diff
@@ -324,21 +324,32 @@ bool IthoCC1101::parseMessageCommand() {
+  const bool isOrconTimer1Command  = checkIthoCommand(&inIthoPacket, orconMessageTimer1CommandBytes);
+  const bool isOrconTimer2Command  = checkIthoCommand(&inIthoPacket, orconMessageTimer2CommandBytes);
+  const bool isOrconTimer3Command  = checkIthoCommand(&inIthoPacket, orconMessageTimer3CommandBytes);
+  const bool isOrconAutoCO2Command = checkIthoCommand(&inIthoPacket, orconMessageAutoCO2CommandBytes);
```
Is it really needed to check all possible types?
It looks like you may be done at some point as I think you will only have one of these set to true, right?

tonhuisman replied:
That part already bothered me when I looked at it, but it's 'inherited' code.

I'll do some improvements there.

TD-er commented:
See: https://github.com/letscontrolit/ESPEasy/blob/84886d05429303b4ecb9608394bb13a87162b5de/lib/Itho/IthoCC1101.cpp#L519-L527
Is this delay needed? Why not re-schedule the task to retry?

TD-er commented:
See: https://github.com/letscontrolit/ESPEasy/blob/84886d05429303b4ecb9608394bb13a87162b5de/lib/Itho/IthoCC1101.h#L93-L95
Please make all get functions const, like this:
```c++
  uint8_t getLastCounter() const {
    return outIthoPacket.counter;
  }
```

TD-er commented:
See: https://github.com/letscontrolit/ESPEasy/blob/84886d05429303b4ecb9608394bb13a87162b5de/lib/Itho/IthoCC1101.h#L16-L48

These are in a global scope, and thus they are created and always present in memory, regardless whether the plugin is being used.
Not sure if these need to be in PROGMEM, as they might be needed while also doing other SPI operations. But there should be a better way to handle these.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[Itho Lib] Code improvements as suggested (mostly)

tonhuisman commented:
Code improvements:
- Move command code arrays to `PROGMEM`
- Avoid collecting all command `bool` vars where only 1 will be true
- Removed (commented out) all unused class methods

Not (yet) changed: The delay on `sendData()` retries, as this involves a major restructure of that part of the code, while other parts of the library might have much longer delays while waiting for the device to respond. Longest delay here will be 3\*40 msec. (or 30\*4 msec. for join command). This only is invoked from `PLUGIN_WRITE`, not from other, more time-critical, `PLUGIN_` events.

- [x] Tested on ESP32

tonhuisman committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring

One35SEven commented:
Hi. I tested on a ESP32S2. Works with ITHO box. Don't see any difference when turning on/off RF Debugging. 
Can't find if this can be used with my Orcon MVS-15RH. Difficult to see if RF packets for Orcon are from remote or fan. 
Only seem to see 1 side.

tonhuisman commented:
> Can't find if this can be used with my Orcon MVS-15RH.

I've found a (Dutch) installation manual for that Orcon unit on the [interwebz](https://www.orcon.nl/producten/mvs-15rh-woonhuisventilator-met-vochtsensor-excl-bediening/), but it only describes how to connect extra remotes, not what codes should be sent, and I haven't found clear instructions in the source code either. There are some groups on Tweakers.net that seem to have some knowledge about this kind of equipment, but I haven't investigated that yest, as I won't be able to verify if that would work.

Maybe the original author of the Orcon part, @nl0pvm , can assist here?
Edit:
You could send the command `state,1111` from the Tools page while the ventilation unit is still in learning mode, that's the way to connect to an ITHO unit, might also work for Orcon.


> Don't see any difference when turning on/off RF Debugging

That should show some more logging in the log output, either via serial or web-log in the UI.
Edit:
It will show log lines starting with `ITHO: Received from ID:` then include the ID from the device that sent it, and the raw command (numeric code) sent.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[P118] Fix LOG_LEVEL_DEBUG compilation

One35SEven commented:
> Maybe the original author of the Orcon part, @nl0pvm , can assist here? Edit: You could send the command `state,1111` from the Tools page while the ventilation unit is still in learning mode, that's the way to connect to an ITHO unit, might also work for Orcon.

Tried state,1111, but doesn't seem to work. Will see if i can find more in the tweakers.net forums. Hopefully @nl0pvm can assist.
> 
> > Don't see any difference when turning on/off RF Debugging
> 
> That should show some more logging in the log output, either via serial or web-log in the UI. Edit: It will show log lines starting with `ITHO: Received from ID:` then include the ID from the device that sent it, and the raw command (numeric code) sent.
With debugging off, i also see the "ITHO: Received from ID: << ID >> messages.

tonhuisman commented:
> With debugging off, i also see the "ITHO: Received from ID: << ID >> messages.

What build are you running, a Release build of ESPEasy, of downloaded from this PR's Actions build? That may be the difference, as I tried to modify that behavior, AFAICS...

One35SEven commented:
> What build are you running, a Release build of ESPEasy, of downloaded from this PR's Actions build? That may be the difference, as I tried to modify that behavior, AFAICS...

Didn't realize there was a new version (not that well versed in github)

Was running:
Build: ESP_Easy_mega_20220824_collection_E_ESP32s2_4M316k Aug 24 2022
Upgraded to last night:
Build: ESP_Easy_mega_20220914_collection_E_ESP32s2_4M316k Sep 14 2022

I don't see a difference when enabling RF debug. 

`
Enable RF DEBUG log: disabled
Enable minimal RF INFO log: disabled
---------------------------------------
3008819: WD   : Uptime 50 ConnectFailures 0 FreeMem 111232 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init
3038819: WD   : Uptime 51 ConnectFailures 0 FreeMem 111380 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init

Enable RF DEBUG log: disabled
Enable minimal RF INFO log: enabled
---------------------------------------
3157320: ITHO: Received from ID: 33,d2,9; raw cmd: 0
3158819: WD   : Uptime 53 ConnectFailures 0 FreeMem 111160 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init
3167040: ITHO: Received from ID: 68,52,15; raw cmd: 0
3177000: ITHO: Received from ID: 78,18,be; raw cmd: 0
3188819: WD   : Uptime 53 ConnectFailures 0 FreeMem 111316 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init
3196661: ITHO: Received from ID: 9a,d2,4; raw cmd: 0

Enable RF DEBUG log: enabled
Enable minimal RF INFO log: enabled
---------------------------------------
3334228: ITHO: Received from ID: 5,26,58; raw cmd: 0
3338820: WD   : Uptime 56 ConnectFailures 0 FreeMem 111436 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init
3346925: ITHO: Received from ID: 82,6a,f2; raw cmd: 0
3368819: WD   : Uptime 56 ConnectFailures 0 FreeMem 111440 WiFiStatus WL_CONNECTED 3 ESPeasy internal wifi status: Conn. IP Init
3368917: ITHO: Received from ID: 78,18,be; raw cmd: 0
3370705: ITHO: Received from ID: 1a,20,14; raw cmd: 0

`

One35SEven commented:
For my Orcon Fan. It seems every once in a while I see the cmd: 101 in the debug log. And the ID that follows that might be the return "OK" from my fan. All not as consistent as for the ITHO remotes. 
But coundn't fill out the hex letters a,b,c,d,f in the device ID fields for the Device ID byte 1, 2 3 fields, only digits and the e.
In my case i needed an a.

tonhuisman commented:
> But coundn't fill out the hex letters a,b,c,d,f in the device ID fields for the Device ID byte 1, 2 3 fields, only digits and the e.
> In my case i needed an a.

Those input fields expect decimal input, but the logged numbers are hexadecimal. You can use the Windows calculator in Developer/Programmer mode to enter each hex value (separately) to see what the decimal value is and fill in those 3 values.

One35SEven commented:
> Those input fields expect decimal input, but the logged numbers are hexadecimal. You can use the Windows calculator in Developer/Programmer mode to enter each hex value (separately) to see what the decimal value is and fill in those 3 values.

Ah; Was thinking about the Unit IDs that are in hex. Thanks will use decimal values for Device ID fields.

tonhuisman commented:
> I don't see a difference when enabling RF debug.

When this option is enabled, and _also_ the Log level is set for Debug, some extra logging is output. This requires the Debug log level to be available, but that's not in the Collection builds...
I'll hide the UI option when it's not useful to be set.

tonhuisman committed:
[P118] Hide Debug log option from UI when not available, small improvements

TD-er committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring

TD-er commented:
This one adds over 5k to the build size.
Seems a lot when just looking at the code.

TD-er commented on lib/Itho/IthoCC1101.cpp:
```diff
@@ -626,11 +754,85 @@ const uint8_t * IthoCC1101::getMessageCommandBytes(IthoCommand command)
+  switch (command)
+  {
+    case IthoStandby:
+      return sizeof(ithoMessageStandByCommandBytes) / sizeof(uint8_t);
```
Not sure if this will be correct.
Please check as I have no idea whether you're checking for the array length or the `sizeof(pointer)`

tonhuisman replied:
Checked a few different sized arrays and it returns the expected size (length of the command bytes array).
Also checked what it could save when replaced by constant values (case with fall-through), but that's already optimized by the compiler (16 bytes gained on ESP32, none on ESP8266), so no reason to bother about that.

TD-er replied:
Hmm that surprises me, since switch statements aren't that "cheap" regarding bin size and also I don't know whether the compiler will already compute the return value at compile time, or that we might need to add `constexpr` to the code to tell the compiler that it can already compute the value.

One35SEven commented:
I have tested with my Orcon FAN and it seems to work. Now trying with 2 CC1101 chips connected to the same ESP32S2. 

One for Itho and one for Orcon fan. After some testing and configuring Home assistant 2 questions:
* How to give command to the 2nd device via HTTP? (http://<IP ESPEASY/control?cmd=State,1 only works for first device)
* It seems that then giving a command via MQTT to the 2nd device (ORCON in my case), the State for the 1st device gets updated in stead of the State for the 2nd device. Might this be a bug?

Also I have some suggestions for the documentation page, where can I put those? Couldn't find where in github.

TD-er commented:
> * How to give command to the 2nd device via HTTP? (http://<IP ESPEASY/control?cmd=State,1 only works for first device)

See [documentation on multiple instances of the same plugin](https://espeasy.readthedocs.io/en/latest/Reference/Command.html?highlight=multiple%20instance#command-a-specific-task-for-multiple-instances-of-a-plugin)

tonhuisman commented:
> Also I have some suggestions for the documentation page, where can I put those? Couldn't find where in github.

You can email that to me (see GH profile), or place your remarks here, and I will include it.

The `docs/source` folder of the repository holds the documentation in ReStructured Text/Sphinx format.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
[Build] Add CLIMATE configuration and move P118 there with all climate sensors

tonhuisman committed:
[Docs] Update documentation for CLIMATE builds, fix some errors in documentation

tonhuisman committed:
[Build] Fix compiler warnings for signed/unsigned int compare

One35SEven commented:
Not sure if I can still report/ask something here, but I 'm running into an issue. Had to wait for new cc1101 chip.
I now have 2 cc1101 chips connected to my esp32. Connected using information on this page on how to connect multiple SPI's.
So 2 tasks configured, one per cc1101 chip. 

Now i'm running into the following. When sending commands via http directly everything works like i would expect.
When sending: http://<IP>/control?cmd=[1].State,102, the state of task 1 gets updated on the devices page.
When sending  http://<IP>/control?cmd=[2].State,2, the state of task 2 gets updated on the devices page.

But when I use MQTT, something is wrong
When I send to the topic for device 1 the state of task 1 gets updated on the devices page.
When i send to the topic for device 2 also the state of task 1 gets updated on the devices page.

So something seems wrong here.

tonhuisman commented:
> But when I use MQTT, something is wrong When I send to the topic for device 1 the state of task 1 gets updated on the devices page. When i send to the topic for device 2 also the state of task 1 gets updated on the devices page.
> 
> So something seems wrong here.

That is most likely not related to this plugin. What is the exact topic you send via MQTT?

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

One35SEven commented:
> > But when I use MQTT, something is wrong When I send to the topic for device 1 the state of task 1 gets updated on the devices page. When i send to the topic for device 2 also the state of task 1 gets updated on the devices page.
> > So something seems wrong here.
> 
> That is most likely not related to this plugin. What is the exact topic you send via MQTT?

esp-fan/FAN01/cmd = State 1
esp-fan/FAN02/cmd = State 101

tonhuisman commented:
Ah, the topic isn't used to address a specific task, as it's not easy to determine what part of the topic relates to a task, if any, so assuming the tasks are named `fan01` and `fan02`, you should use `fan01.state,1` and `fan02.state,101` for the values, similar to the http commands (though you used the tasknumbers there, you can also use the _taskname_ as a prefix. The square braces are optional)

One35SEven commented:
> Ah, the topic isn't used to address a specific task, as it's not easy to determine what part of the topic relates to a task, if any, so assuming the tasks are named `fan01` and `fan02`, you should use `fan01.state,1` and `fan02.state,101` for the values, similar to the http commands (though you used the tasknumbers there, you can also use the _taskname_ as a prefix. The square braces are optional)

Thanks. This works. Although didn't expect to be able to send commands to either cmd-topic.
So sending a value for FAN01 to the topic for FAN02 and vice versa works. Below gets same result:
esp-fan/FAN01/cmd = FAN02.state,100
esp-fan/FAN02/cmd = FAN02.state,100

Still working on right Home assistant configuration not happy with it yet), but when i have everything complete will send you a mail with updates for the documentation. 
(Home Assistant it seems can't handle a fan without an off setting, the templates seem to be for a room fan in stead of a central ventilation system).

TD-er commented:
@One35SEven Have you seen this section in the docs? [Home Assistant (openHAB) MQTT - Command Handling](https://espeasy.readthedocs.io/en/latest/Controller/C005.html#command-handling)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P118-select-cs-and-refactoring

tonhuisman committed:
Merge branch 'mega' into feature/P118-select-cs-and-refactoring
